fix: Changed edit profile's link based social system

scixor opened the pull request:
- Fixes #2591 
- Fixes #2590 

## Changes:
- changed the way users add the links to use usernames instead of adding an entire link
- now making changes to Instagram link works (it was a one line change)
- this doesn't change the way we interface from the server but we just hides it from the User

## How this is done
- when we retrieve link from the server to display we strip off the protocol part by using this regEx pattern `((https)|(http)|[/|:]|(.com)|(www.)|(instagram)|(facebook)|(twitter))` which just leaves the username part which is then shown to the screen
- when user clicks on update button the input is taken from the fields and we run it through another regEx pattern `[$\s@,]` to remove any sort of extra things like space @ , ... and then add it to the back of a header URL.

you can check that expression here by pasting into https://www.regextester.com/ the regular expression box and changing the input text

Video for the change:
https://photos.app.goo.gl/FWfMLu377WTStAbSA

APK:
[app-fdroid-debug.zip](https://github.com/fossasia/open-event-attendee-android/files/4099766/app-fdroid-debug.zip)

iamareebjamal commented on app/src/main/res/values/strings.xml:
```diff
@@ -432,7 +432,10 @@
+    <string name="invalid_url_message">Please enter the correct username</string>
+    <string name="facebookHeadUrl">https://www.facebook.com/</string>
+    <string name="twitterHeadUrl">https://www.twitter.com/</string>
+    <string name="instagramHeadUrl">https://www.instagram.com/</string>
```
These are untranslatable strings and shouldn't be present here

scixor replied:
where should I keep them then?

scixor replied:
I am adding `translatable="false"` tag to them and moving them at the bottom

iamareebjamal commented:
Add unit tests for extraction logic

scixor commented:
@iamareebjamal there seems to be a bigger issue... can you please check this with the play store version about this... when you add a social media link.. you can't remove it later.. I was thinking it was my fault from last 3 hours.. seems playstore version also has this

scixor commented:
Also I have added a video in the description please check that too

scixor commented:
Added tests for regEx patterns

iamareebjamal commented on app/src/main/res/values/strings.xml:
```diff
@@ -535,4 +535,8 @@
+    <!-- Non-Translatable -->
+    <string name="facebookHeadUrl" translatable="false">https://www.facebook.com/</string>
+    <string name="twitterHeadUrl" translatable="false">https://www.twitter.com/</string>
+    <string name="instagramHeadUrl" translatable="false">https://www.instagram.com/</string>
```
snake case

scixor replied:
oh.. I thought we used Camel case.. I will fix this

iamareebjamal commented:
We want to show error and not silently convert usernames with spaces to without them

scixor commented:
i think we can add a warning saying... spaces will be removed automatically when updated or something? would that work?

scixor commented:
Done all the changes.. I have added a text view which pops up slowly mentioning this .. and also changed Update button to different color ... for Identifying that there might be an issue... I didn't add red because it didn't match with our cool bluish theme, and also this is just a warning.. 😀

scixor commented:
@iamareebjamal @liveHarshit linting is now also fixed 😀builds should pass now in Travis

liveHarshit commented on app/src/main/java/org/fossasia/openevent/general/auth/EditProfileFragment.kt:
```diff
@@ -32,17 +32,7 @@ import kotlinx.android.synthetic.main.dialog_edit_profile_image.view.editImage
-import kotlinx.android.synthetic.main.fragment_edit_profile.view.toolbar
-import kotlinx.android.synthetic.main.fragment_edit_profile.view.twitter
-import kotlinx.android.synthetic.main.fragment_edit_profile.view.updateButton
+import kotlinx.android.synthetic.main.fragment_edit_profile.view.*
```
Remove wildcard import.

scixor replied:
I will fix that

liveHarshit commented on app/src/test/java/org/fossasia/openevent/general/RegExTest.kt:
```diff
@@ -0,0 +1,18 @@
+
+    @Test
+    fun `anomaly removal when updating`() {
+        assertEquals("helloworld", frag.updateStringRegEx("h@ello@wo ,   rld"))
```
it's invalid!

scixor replied:
@liveHarshit  why? I am new to the tests 😇 can you please explain?

scixor replied:
it tests the function's use of removing the anomalies... I don't get the problem 😅

scixor replied:
I can't use that function to make entire links because .. it is also being used by change listener to analyze whether the text is right or not...

scixor commented:
@liveHarshit made the changes asked for.. other than test part.

iamareebjamal commented on app/src/main/res/layout/fragment_edit_profile.xml:
```diff
@@ -152,13 +154,26 @@
+            android:padding="@dimen/padding_medium"
+            android:visibility="gone"
+            android:textAlignment="center"
+            android:text="@string/error_text" />
```
Remove this and use TextInput error

scixor replied:
sure

iamareebjamal commented on app/src/test/java/org/fossasia/openevent/general/RegExTest.kt:
```diff
@@ -0,0 +1,18 @@
+
+    @Test
+    fun `anomaly removal when updating`() {
+        assertEquals("helloworld", frag.updateStringRegEx("h@ello@wo ,   rld"))
```
Since this is invalid username, it should throw. We are only going to make benign changes like URLs to username conversions and removing of `@` sign and it should be visible to user before they click save

scixor replied:
but invalid username is confirmed at a different function 😅 that function decides if the username is invalid or not.. this is just a regEx pattern function which takes a string removes any anomaly and return a string **(tests if regEx pattern removes any essential things)** .. I will make a new test case for invalid username

iamareebjamal replied:
It should not do this. This is an invalid conversion

scixor replied:
@iamareebjamal so how do we check and change the UI with onChangeListner and remove things when user Clicked on update with a same function 😕 ? this function is being used when updating and also while checking. I added that test if someone changes the pattern and if it leads to deleting of wrong characters .. now I also added another test...

scixor replied:
I think this solves this conversation

scixor committed:
Added regular expression tests

scixor committed:
Changed profile edit's socials to username based

Sid911 commented:
Fixed most of the things.. 😀

scixor committed:
Merge branch 'development' into development

scixor committed:
Merge branch 'development' into development

scixor committed:
Merge branch 'development' into development

scixor commented:
well if someone gets time.. please review this @iamareebjamal @liveHarshit @nikit19

scixor committed:
Merge branch 'development' into development

scixor committed:
Merge branch 'development' into development

scixor commented:
@liveHarshit please review 😃... thanks

scixor committed:
Merge branch 'development' into development

liveHarshit committed:
Merge branch 'development' into development

scixor commented:
closing pr! because of no response whatsoever on this pr by maintainers 🤷‍♂️.. Maybe someone else could fix it.
